src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnelSources/OutlineTunnel.swift

```diff
@@ -13,54 +13,156 @@
+        ipv4Settings.includedRoutes = [NEIPv4Route.default()]
+        ipv4Settings.excludedRoutes = getExcludedIpv4Routes()
+        settings.ipv4Settings = ipv4Settings
 
-  public convenience init(id: String, config: [String: Any]) {
-    self.init()
-    self.id = id
-    self.host = config["host"] as? String
-    self.password = config["password"] as? String
-    self.method = config["method"] as? String
-    if let port = config["port"] {
-      self.port = String(describing: port)  // Handle numeric values
+        // Configure with Cloudflare, Quad9, and OpenDNS resolver addresses.
+        settings.dnsSettings = NEDNSSettings(servers: ["1.1.1.1", "9.9.9.9", "208.67.222.222", "208.67.220.220"])
+        
+        return settings
     }
-    if let prefix = config["prefix"] as? String {
-      self.prefix = Data(prefix.utf16.map{UInt8($0)})
+}
+
+// Returns all IPv4 addresses of all interfaces.
+func getNetworkInterfaceAddresses() -> [String] {
+    var interfaces: UnsafeMutablePointer<ifaddrs>?
+    var addresses = [String]()
+    
+    if getifaddrs(&interfaces) != 0 {
+        NSLog("Failed to retrieve network interface addresses")
+        return addresses
     }
-  }
+    
+    var interface = interfaces
+    while interface != nil {
+        // Only consider IPv4 interfaces.
+        if interface?.pointee.ifa_addr.pointee.sa_family == UInt8(AF_INET) {
+            let addr = interface!.pointee.ifa_addr!.withMemoryRebound(to: sockaddr_in.self, capacity: 1) { $0.pointee.sin_addr }
+            if let ip = String(cString: inet_ntoa(addr), encoding: .ascii) {
+                addresses.append(ip)
+            }
+        }
+        interface = interface?.pointee.ifa_next
+    }
+    
+    freeifaddrs(interfaces)
+    
+    return addresses
+}
 
-  public func encode() -> Data? {
-    return try? JSONEncoder().encode(self)
-  }
+let kVpnSubnetCandidates: [String: String] = [
+    "10": "10.111.222.0",
+    "172": "172.16.9.1",
+    "192": "192.168.20.1",
+    "169": "169.254.19.0"
+]
 
-  public static func decode(_ jsonData: Data) -> OutlineTunnel? {
-    return try? JSONDecoder().decode(OutlineTunnel.self, from: jsonData)
-  }
+// Given the list of known interface addresses, returns a local network IP address to use for the VPN.
+func selectVpnAddress(interfaceAddresses: [String]) -> String {
+    var candidates = kVpnSubnetCandidates
+    
+    for address in interfaceAddresses {
+        for subnetPrefix in kVpnSubnetCandidates.keys {
+            if address.hasPrefix(subnetPrefix) {
+                // The subnet (not necessarily the address) is in use, remove it from our list.
+                candidates.removeValue(forKey: subnetPrefix)
+            }
+        }
+    }
+    if candidates.isEmpty {
+        // Even though there is an interface bound to the subnet candidates, the collision probability
+        // with an actual address is low.
+        return kVpnSubnetCandidates.randomElement()?.value ?? ""
```

Comment:
I think `randomElement()` must not be `nil` here? `randomElement()!.value` ?

Reply:
It can be nil if the collection is empty.

I changed the code a bit to use your suggestion and make it clearer that candidates is not empty.